diff --git a/liblp/utility.cpp b/liblp/utility.cpp
index 4f7b0ce..1866233 100644
--- a/liblp/utility.cpp
+++ b/liblp/utility.cpp
@@ -34,6 +34,17 @@
 
 #include "utility.h"
 
+#ifndef TEMP_FAILURE_RETRY
+#define TEMP_FAILURE_RETRY(expression)                 \
+  (__extension__({                                    \
+    long int __result;                                \
+    do {                                               \
+      __result = (long int)(expression);               \
+    } while (__result == -1L && errno == EINTR);      \
+    __result;                                          \
+  }))
+#endif
+
 namespace android {
 namespace fs_mgr {
 
